Bg schema

rauf543 opened the pull request:
# Pull Request
QUICK EDIT: First time doing a PR, so if something is wrong for the format, or with the code, or with the way I have committed etc... please let me know and I will try my best to fix it
Overview

This pull request introduces a new feature to dynamically manage combat background schemas for events and other use cases without requiring hardcoding. It enables admins and event/content staff to add, edit, delete, and activate background schemas through a front-end interface, with changes seamlessly reflected in the combat system.

Back-End Changes

Updated determineArenaBackground and determineCombatBackground:

Replaced constant-based logic with dynamic queries to a new database table: backgroundSchemas.
Functions now query the table to find the active schema and fetch its schema field, which redirects to the appropriate link.
Once fetched, the battle logic remains unchanged.
Database Integration:

Added the backgroundSchemas table to store schema data, including an active status flag.
Supports CRUD operations and status toggling for dynamic updates.
Refactored Function Calls:

Updated all existing calls to determineArenaBackground and determineCombatBackground to align with the new dynamic functionality.
Front-End Changes

New Admin Page:
![1](https://github.com/user-attachments/assets/e2f8cdf1-6403-4075-aa85-edd99937bd4c)
[QA Testing Checklist.pdf](https://github.com/user-attachments/files/17893294/QA.Testing.Checklist.pdf)
![2](https://github.com/user-attachments/assets/b1099f5e-08fa-4962-9c14-503cca6abe84)
![3](https://github.com/user-attachments/assets/8d4d6669-3449-4c7c-be25-fe2fd77ec711)
![4](https://github.com/user-attachments/assets/bd204ba9-792f-4d35-870c-0e48ebdff0bc)

Added a page under the Manuals > Combat section called Background Schemas.
Functionality includes:
Adding new background schemas.
Editing existing schemas.
Deleting unwanted schemas.
Setting active schemas for immediate use in events.
Permission Locking:

Restricted access to this page based on specific permissions.
Permissions required: •	

User: No Access  ✔
Coding Admin: Full CRUD access with ability to change active schema. ✔
Content Admin: Full CRUD access with ability to change active schema. ✔
Moderator Admin: Limited CRUD Access, no ability to change active Schema or delete schemas. ✔
Head Moderator/Moderator: No Access ✔
Content/Event: Limited CRUD Access, no ability to change active Schema or delete schemas. ✔

Summary
This change provides greater flexibility and efficiency for managing combat backgrounds, especially for dynamic events. It simplifies staff workflows by eliminating the need for code-level changes while maintaining secure access control.
## License

By making this pull request, I confirm that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the Studie-Tech ApS organization has the copyright to use and modify my contribution for perpetuity.

MathiasGruber commented:
First impression is this looks awesome!.. I will review it as soon as possible - I'm down with a cold these days, so my head is not clear enough right now, but will prioritize this as soon as I can! Once merged, this definitely rewards a good chunk of reputation points - which account would you like these to go to?

MathiasGruber commented on app/src/validators/backgroundSchema.ts:
```diff
@@ -0,0 +1,23 @@
+  name: z.string().min(1).max(191),
+  description: z.string().min(1).max(191),
+  isActive: z.boolean().default(false),
+  schema: z.object({
```
Extract `schema` to a separate validator, which is also exported and used here, i.e. 

```
export const bgSchema = z.object({
    ocean: z.string().url(),
    ice: z.string().url(),
    dessert: z.string().url(),
    ground: z.string().url(),
    arena: z.string().url(),
    default: z.string().url(),
  })```
  
  Then we can use this validator & its inferred type in the `schema.ts` file

MathiasGruber commented on app/drizzle/schema.ts:
```diff
@@ -2260,3 +2260,39 @@ export const userRequestRelations = relations(userRequest, ({ one }) => ({
+      .notNull()
+      .default(sql`(UUID())`),
+    schema: json("schema")
+      .$type<{
```
Reuse the inferred type from the validated file - so that we only need to maintain it one place :)

MathiasGruber commented on app/pnpm-lock.yaml:
Why did the lock file change if we don't have new packages?

MathiasGruber commented on app/src/app/manual/combat/backgroundSchema/edit/[schemaId]/page.tsx:
```diff
@@ -0,0 +1,220 @@
+}) => {
+  const router = useRouter();
+  const utils = api.useUtils();
+  const [imageUrls, setImageUrls] = useState<Record<string, string>>({
```
Maybe also re-use from the validators file, if we add default values of "" to the zod schema? Might not work, not sure.

MathiasGruber commented on app/src/server/api/routers/backgroundSchema.ts:
```diff
@@ -0,0 +1,133 @@
+    let newName = baseName;
+    let suffix = 1;
+    const id = nanoid();
+
```
Permission check missing?

MathiasGruber commented on app/src/server/api/routers/backgroundSchema.ts:
```diff
@@ -0,0 +1,133 @@
+      }),
+    )
+    .mutation(async ({ ctx, input }) => {
+      const { id, ...data } = input;
```
Permission check missing?

MathiasGruber commented on app/src/server/api/routers/backgroundSchema.ts:
```diff
@@ -0,0 +1,133 @@
+  delete: protectedProcedure
+    .input(z.object({ id: z.string() }))
+    .mutation(async ({ ctx, input }) => {
+      await ctx.drizzle
```
Permission check missing?

MathiasGruber commented on app/src/server/api/routers/backgroundSchema.ts:
```diff
@@ -0,0 +1,133 @@
+    .input(z.object({ id: z.string() }))
+    .mutation(async ({ ctx, input }) => {
+      const { id } = input;
+
```
Permission check missing?

MathiasGruber commented on app/src/server/api/routers/combat.ts:
```diff
@@ -552,6 +543,7 @@ export const combatRouter = createTRPCRouter({
       }
       // Determine battle background
       if (selectedAI) {
+        const background = await determineArenaBackground(ctx.drizzle, "arena");
```
I'm wondering if we can figure out a way in which the queries for fetching background & user information (within initiateBattle) can be run in parallel within Promise.all.. every time we introduce sequential queries, we increase the function call time significantly, and the intiate battle call is one which is quite frequent.

MathiasGruber commented on das:
Don't think this should be comitted? :)

MathiasGruber reviewed (dismissed):
Overall super good work, and very awesome feature.. I've not tried spinning it up, but I thought all this work deserved a swift review.. I found a few places where I think we could make it better - specifically in trying to have a one source of truth for the schema and validators, and then also performance in terms of running parallel instead of sequential queries during battle initiation

rauf543 commented:
Thank you for the kind words! I am still not too experienced with this technology so appreciate all of the comments. Ill work on the raised issues, most seem pretty straightforward to resolve.

rauf543 commented:
Pushed requested changes, adjusted initiateBattle to work with a parallel promise, added permissions checks for the schema router aswell as tried to use one source of truth (validator) for all schema checks.

rauf543 commented:
> Once merged, this definitely rewards a good chunk of reputation points - which account would you like these to go to?

Thank you for the unexpected but welcome reward! my ign is Dirtyhands.

rauf543 commented:
edge case bug identified.

When you edit a schema, it kicks you back to the getAll page, if you press edit again, it shows you the previous, unedited fields. Probably some other state I need to invalidate. Going to bed right now so will check more on this tommorow.

The bug is minor and after a refresh all is how it should be.

MathiasGruber commented:
Will review this tonight when I have time! 🤘

MathiasGruber commented:
I made a mistake and merged this prematurely.. don't worry, I'll fix up the errors and get it working again. Will still upload reputation points to your account once I have it fixed up :)

rauf543 commented:
Not a problem, don’t worry too much about the reps, I do it to learn and for fun, your guidance along the way is much more valuable, thank you for that.

I just wanted to ask, what was the problem? Just so I make sure to commit better code next time so it can be merged straight away.

MathiasGruber commented:
Just a lot of type errors on the build.. if you run a `make pnpm build` it'll try to build locally, which should reveal all missing type errors :)

MathiasGruber commented:
Here's the fix I applied to make it compile (and a bit of UX alignment) https://github.com/MathiasGruber/TheNinjaRPG/commit/7c72ab41e75c79e1c2fcff895da1f04041b9e254

Will send 200 reputation points to your account, thanks for the contribution :)
